### PR TITLE
[WIP] short srtp constrain syntax / implementing "inline interfaces"

### PR DESCRIPTION
**Pull request by realvictorprm**

### CODE REVIEW

**Comment by realvictorprm** [comment]
This PR provides the feature to shorten SRTP constrain syntax with a so called "Inline Interface". Those interfaces should only be target for SRTP's. Next I'm going to implement / remove the requirement of using the SRTP call syntax if the constrains have been defined over such an inline interface.

See the picture below what it spites out right now:
image

**Comment by realvictorprm** [comment]
The short SRTP syntax isn't working fully yet but it's nearly there 😄 
image

**Comment by realvictorprm** [comment]
The trait stuff of the compiler is a bit awful, only through debugging it was possible for me to understand what the hell I need to hand over to particular functions. This really needs refactoring.

**Comment by abelbraaksma** [comment]
This looks very interesting and promising. Is it related to type classes? I use SRTP a lot but it often gets unwieldy, wonder if this alleviates some of the burden of writing SRTP. Is it related to an rfc?

**Comment by jwosty** [comment]
Hmm... are inline interfaces even really necessary if the simpler STRP syntax exists? Isn't the pain that they solve alleviated? Good work, though.

**Comment by realvictorprm** [comment]
@jwosty I think that inline interface are still good for situation when you have too many constrains which would make the declaration unreadable.

**Comment by jwosty** [comment]
@realvictorprm I can see that.

**Comment by TIHan** [comment]
This is interesting. Do we have a RFC?

**Comment by realvictorprm** [comment]
@TIHan not yet but if you want one I'll make one.

**Comment by PatrickMcDonald** [comment]
It looks like the RFC is here: [https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1024-simplify-call-syntax-for-statically-resolved-member-constraints.md]

**Comment by realvictorprm** [comment]
@PatrickMcDonald I thought he was speaking about the inline interfaces.

**Comment by gusty** [comment]
> I thought he was speaking about the inline interfaces.

I think this is confusing, because:

 - You are working on many related features at the same time, although there's nothing wrong with it.
 - Some stuff is not RFCed or not even approved, which is fine but that means that there is a risk of not getting the stuff merged in the end,
 - This PR mentions inline interfaces, why don't we work on that RFC where it applies also to existing trait calls and update this PRs title?

**Comment by realvictorprm** [comment]
The reason why I'm mixing features is because I think we should bunch a few of them together so that:
1. They are released together and not in pieces.
2. The workflow is already guaranteed to work.
3. The features can be build together to ensure that the code is cleaner.

Besides this PR was a proof of concept to show that it's possible to get started with adding the features within a few days of work.

And I'm certainly planning to work on the existing PR's next, now I'm better into the topic after I've played around with this compiler area.

**Comment by cartermp** [comment]
@realvictorprm There's no issue with doing this per say, but if one feature is accepted and another is rejected then you'll have to refactor - it might be easier to do one PR per feature in the future.

I will ask that you track in the PR text itself the features you are implementing and link back to each language design issue. If everything were to be approved, we'd also require an RFC for each feature and a link back to those RFCs in this PR as well.

**Comment by realvictorprm** [comment]
Well OK @cartermp
I'm going to move the simple syntax out of this repository and make a separate PR during trying to get the inline interfaces approved.

File: src/fsharp/IlxGen.fs [hunk 1]
```diff
@@ -6270,11 +6270,11 @@ and GenTypeDef cenv mgbuf lazyInitInfo eenv m (tycon:Tycon) =
            match  tyconRepr with 
            | TFSharpObjectRepr o -> 
                match o.fsobjmodel_kind with 
-               | TTyconClass      -> ILTypeDefKind.Class
-               | TTyconStruct     -> ILTypeDefKind.ValueType
-               | TTyconInterface  -> ILTypeDefKind.Interface
-               | TTyconEnum       -> ILTypeDefKind.Enum
-               | TTyconDelegate _ -> ILTypeDefKind.Delegate
+               | TTyconClass       -> ILTypeDefKind.Class
```
Comment (dsyme) [inline src/fsharp/IlxGen.fs]: Please remove all whitespace changes like this, thanks, it helps see the real diff

File: src/fsharp/TypeChecker.fs [hunk 2]
```diff
@@ -4224,10 +4224,55 @@ let rec TcTyparConstraint ridx cenv newOk checkCxs occ (env: TcEnv) tpenv c =
     | WhereTyparSubtypeOfType(tp, ty, m) ->
         let ty', tpenv = TcTypeAndRecover cenv newOk checkCxs ItemOccurence.UseInType env tpenv ty
         let tp', tpenv = TcTypar cenv env newOk tpenv tp
-        if newOk = NoNewTypars && isSealedTy cenv.g ty' then 
-            errorR(Error(FSComp.SR.tcInvalidConstraintTypeSealed(), m))
-        AddCxTypeMustSubsumeType ContextInfo.NoContext env.DisplayEnv cenv.css m NoTrace ty' (mkTyparTy tp') 
+        let tryGetInlineInterface =
+            match ty' with
```
Comment (dsyme) [inline src/fsharp/TypeChecker.fs]: You should always use the helpers to deconstruct, especially `tryDestAppTy` in this case. This look through type inference variables and type abbreviations

File: src/fsharp/TypeChecker.fs [hunk 3]
```diff
@@ -4224,10 +4224,55 @@ let rec TcTyparConstraint ridx cenv newOk checkCxs occ (env: TcEnv) tpenv c =
...
+            | TType_app(possibleInterface, _) ->
+                match possibleInterface.TryDeref with
+                | VSome interface' when interface'.IsFSharpInlineInterfaceTycon -> 
+                    Some interface'
+                | _ -> None
+            | _ -> None
+        match tryGetInlineInterface with
+        | Some interface' ->
+            let rec collectArgumentTypes ttype args =
+                match ttype with
```
Comment (dsyme) [inline src/fsharp/TypeChecker.fs]: Likewise here. Basically never match directly on TType

File: src/fsharp/TypeChecker.fs [hunk 4]
```diff
@@ -4224,10 +4224,55 @@ let rec TcTyparConstraint ridx cenv newOk checkCxs occ (env: TcEnv) tpenv c =
...
+                    let args1 = collectArgumentTypes arg1 []
+                    args0 @ args1
+                | TType_tuple(_, args) ->
+                    args |> List.collect (fun arg -> collectArgumentTypes arg [])
+                | TType_app (_,_) as r -> args @ [r]
+                | TType_var _ as r -> args @ [r]
+                | _ -> args
+            for member' in interface'.MembersOfFSharpTyconSorted do
+                match member'.Type with
+                | TType_fun(_, args)->
```
Comment (dsyme) [inline src/fsharp/TypeChecker.fs]: Likewise here.  (I'm not totally sure what this is doing as I'm taking a quick glance but I suspect there are existing helpers to do the sort of thing you're doing here)

Comment (realvictorprm) [reply src/fsharp/TypeChecker.fs]: I'll have a look whether there's a fitting one.

**Comment by zpodlovics** [comment]
Looks really promising, what is the status of this patch? It has a potential to speed up the language significantly, because the typical indirect call overhead is high (however recently coreclr have some nice devirtualization improvement) : https://github.com/Microsoft/visualfsharp/issues/3737#issuecomment-385038400

**Comment by realvictorprm** [comment]
This was kind of a draft and experiment which will have to wait until certain SRTP bugs are fixed. I'm going to start an extra thread about SRTP fix on the Fsharp forum.

**Comment by zpodlovics** [comment]
Still looks probably the best way to deal with SRTPs, any update?

**Comment by realvictorprm** [comment]
Have a look onto my post on the F# forums :)

**Comment by zpodlovics** [comment]
For future reference, the F# Forum Post is here: https://forums.fsharp.org/t/srtps-bugs-and-future-plans/96

**Comment by realvictorprm** [comment]
I'm not interested in this right anymore and don't want to leave wrong hopes therefor I'll close. Sorry.

**Comment by NinoFloris** [comment]
@realvictorprm what happened?

**Comment by DemuirGos** [comment]
with F# 6 released and having the philosophy of simplicity and being dev friendly I'd suspect this feature was included in it, so I came to ask what happened to this ?

**Comment by cartermp** [comment]
@DemuirGos it was closed 🙂

**Pull request closed without merging**